zebra/zserv.c

```diff
@@ -524,6 +552,9 @@ static void zserv_process_messages(struct event *thread)
 	/* Reschedule ourselves if necessary */
 	if (need_resched)
 		zserv_event(client, ZSERV_PROCESS_MESSAGES);
+
+	/* Signal a wake up to client IO thread */
+	zserv_client_event(client, ZSERV_CLIENT_READ);
```

Comment:
In an earlier version, I think there was a condition that there was actually room in the incoming queue/fifo; I think that's a good test to make.

Reply:
Earlier version condition signals client IO only if the (ibuf_count == p2p). But we have a bug here. 
 - Imagine current p2p =1000. Client IO has filled 1000 items in ibuf_fifo and is waiting for the main thread to signal. 
 - However, before the main thread wakes up if p2p is configured to say 2000 and then the main wakes up, it sees ibuf_cnt(1000) is not equal to p2p(2000), and the read thread never gets woken up. 

Essentially the older version has an issue handling the p2p config changes.

Reply:
I agree that there were bugs in older version, but I'm not sure I understood that comment - do you agree that there should be a test that there is room in the fifo? I believe that test a) could be made correctly, and b) would be more clear and efficient.

Reply:
I don't agree here because with the current code in review logic, 

- Client IO reschedules itself if max limit is NOT reached
- Main thread reschedules itself if fifo has more items, and it signal client IO thread always (ensures to include the read socket in the select/poll/etc).

The current code handles all conditions wherein it read thread(client io) tries to always add more items and main thread always tries to process the item. 

Older code has a timing issue where client IO never wakes up because if main thread wakes up after the config change (limit2>limit1), it sees that the fifo count is not same as p2p and never signals client IO. 

I have tested this portion where i perform shutdown/no shutdown in a loop and in between change the configs (increase and decrease p2p) and it works fine..

Reply:
So my concern is this: currently, we have just one "value", one limit, that we're applying. but it's not uncommon in flow-control to have more patterns, where we might for example accept a burst, and then defer reads until the buffer occupancy reaches some "low-water" level. if we have a test that schedules the read event only when there's reason to believe that there is room in the buffer (the inbound queue), that will not have to change if the mechanism evolves further. It's always correct to say "allow reads if there's room in the inbound queue", whatever else happens, but it may not always be correct to schedule reads constantly.

Reply:
only case where main knows there is not enough space is when 
- client IO filled its max (say 1000)
- user changed conf to 400
- now main wakes and sees it still has 600 more items and there is no space in the buffer.. 

so a condition like 
after processing , if fifo head exists and fifo-count >p2p, do nothing.. else wake the read thread.. 

does this work?

Reply:
No, I think I'm back to "simple" - I'll resolve this comment, and we can move on.